velox/connectors/hive/PartitionIdGenerator.cpp

```diff
@@ -0,0 +1,139 @@
+      maxId_(0) {
+  hashers_.reserve(partitionChannels_.size());
+  for (column_index_t i = 0; i < partitionChannels_.size(); i++) {
+    hashers_.push_back(exec::VectorHasher::create(
+        inputType->childAt(partitionChannels_[i]), partitionChannels_[i]));
+  }
+}
+
+std::map<uint64_t, uint64_t> PartitionIdGenerator::run(
+    const RowVectorPtr& input,
+    raw_vector<uint64_t>& result) {
+  result.resize(input->size());
+  bool rehashed =
+      computeIds(input, SelectivityVector{input->size(), true}, result);
+
+  std::map<uint64_t, uint64_t> idMap;
+  if (rehashed && allPartitionsVector_->size() > 0) {
+    raw_vector<uint64_t> newIds(allPartitionsVector_->size());
+
+    VELOX_CHECK(
+        !computeIds(allPartitionsVector_, activeIds_, newIds),
+        "Partition ID generator should not rehash when computing for the existing partitions.");
+
+    activeIds_.applyToSelected([&](auto id) { idMap[id] = newIds[id]; });
+
+    movePartitions(idMap);
+  }
+
+  addPartitions(input, result);
+
+  return idMap;
+}
+
+bool PartitionIdGenerator::computeIds(
+    const RowVectorPtr& input,
+    const SelectivityVector& rows,
+    raw_vector<uint64_t>& result) {
+  result.resize(input->size());
+
+  for (auto& hasher : hashers_) {
+    auto partitionVector = input->childAt(hasher->channel())->loadedVector();
+    hasher->decode(*partitionVector, rows);
+  }
+
+  bool rehashed = false, toRehash = false;
+  do {
+    toRehash = false;
+    for (auto& hasher : hashers_) {
+      if (!hasher->computeValueIds(rows, result)) {
+        toRehash = true;
+      }
+    }
+    if (toRehash) {
+      rehashed = true;
+
+      uint64_t multiplier = 1;
+      for (auto i = 0; i < hashers_.size(); i++) {
+        multiplier = hashers_[i]->enableValueIds(multiplier, kHasherReservePct);
+        VELOX_CHECK_NE(
+            multiplier,
+            exec::VectorHasher::kRangeTooLarge,
+            "Exceeded limit of distinct partitions.");
+      }
+    }
+  } while (toRehash);
+
+  return rehashed;
+}
+
+void PartitionIdGenerator::movePartitions(
+    const std::map<uint64_t, uint64_t>& indexMap) {
+  vector_size_t maxNewId = indexMap.rbegin()->second;
+
+  allPartitionsVector_->resize(maxNewId + 1, false);
+  activeIds_.resize(maxNewId + 1, false);
+
+  for (auto entry = indexMap.rbegin(); entry != indexMap.rend(); entry++) {
+    allPartitionsVector_->copy(
+        allPartitionsVector_.get(), entry->second, entry->first, 1);
+    allPartitionsVector_->setNull(entry->first, true);
+    activeIds_.setValid(entry->second, true);
+    activeIds_.setValid(entry->first, false);
+  }
+  activeIds_.updateBounds();
+}
+
+void PartitionIdGenerator::addPartitions(
+    const RowVectorPtr& input,
+    const raw_vector<uint64_t>& inputIds) {
+  maxId_ = *std::max_element(inputIds.begin(), inputIds.end());
+  if (maxId_ > allPartitionsVector_->size() - 1) {
+    allPartitionsVector_->ensureWritable(SelectivityVector::empty(maxId_ + 1));
```

Comment:
Why do we need ensureWritable here? Can't we just call resize?

Reply:
Calling only `allPartitionsVector_.resize()` here hits the error

Line: /Users/gaoge/Program/presto/presto-native-execution/velox/velox/vector/FlatVector-inl.h:227, Function:copyValuesAndNulls, Expression: BaseVector::length_ >= targetIndex + count (0 vs. 1203), Source: RUNTIME, ErrorCode: INVALID_STATE

when running the lines

>       for (vector_size_t channel : partitionChannels_) {
>         allPartitionsVector_->childAt(channel)->copy(
>             input->childAt(channel)->loadedVector(), id, row, 1);
>       }


I got ensureWritable() from the body of `RowVector::copy()`, which does more than `RowVector::resize()`.

Reply:
I see. Resizing RowVector requires calling resize on the RowVector and all its children:

```
    allPartitionsVector_->resize(maxId_ + 1);
    for (auto& child : allPartitionsVector_->children()) {
      child->resize(maxId_ + 1);
    }
```

Using ensureWritable here is confusing since the vector is owned by the ID generator, hence, it should always be writable.